api/src/legacy_soap_api/applicants/services/get_opportunity_list_response.py

```diff
@@ -38,6 +49,43 @@ def get_competitions_from_opportunity_list_request(
     )
 
 
+def get_tinstructions_map(db_session: db.Session, competitions: list) -> dict:
+    """Get instructions extension for competitions
+
+    This method returns a map of extension string from the staging.tinstructions table
+    that is needed to build the instructions_url property of the competition for the
+    GetOpportunityListResponse.
+
+    This will return a map of Competition.legacy_competition_id: staging.tinstructions.extension.
+    """
+    legacy_competition_ids = [
+        c.legacy_competition_id for c in competitions if c.legacy_competition_id is not None
+    ]
+    tinstructions = (
+        db_session.execute(
+            select(Tinstructions.extension, Tinstructions.comp_id).where(
+                Tinstructions.comp_id.in_(legacy_competition_ids)
+            )
+        )
+        .mappings()
+        .all()
+    )
+    if not tinstructions:
+        return {}
+    return {tinstruction.comp_id: tinstruction.extension for tinstruction in tinstructions}
+
+
+def get_tinstructions_urls(competition: Competition, tinstructions_map: dict) -> TinstructionsURLs:
+    urls = TinstructionsURLs()
+    if not competition.legacy_package_id:
+        return urls
+    base_url = f"{get_soap_config().grants_gov_uri}/apply/opportunities"
+    urls.schema_url = f"{base_url}/schemas/applicant/{competition.legacy_package_id}.xsd"
+    if extension := tinstructions_map.get(competition.legacy_competition_id):
+        urls.instructions_url = f"{base_url}/instructions/{competition.legacy_package_id}-instructions{ensure_dot_prefix(extension)}"
```

Comment:
For building path strings, could you use `file_util.join()` - it _should_ handle any issues with having an extra slash in one of the strings, just in case.